File: src/GameState.cs
using ElevenLegends.Competition;
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;
using ElevenLegends.Economy;
using ElevenLegends.Manager;
using ElevenLegends.Simulation;
using ElevenLegends.Transfers;

namespace ElevenLegends;

/// <summary>
/// Central game state. Tracks the current day, clubs, competitions, and manager.
/// Call AdvanceDay() for auto-mode, or use PrepareMatchDay/FinishDay for interactive mode.
/// </summary>
public sealed class GameState
{
    private readonly List<Club> _clubs;
    private readonly CompetitionManager _competition;
    private readonly List<SeasonDay> _calendar;
    private readonly int _baseSeed;

    private int _currentDayIndex;
    private int _nationalMatchDayCount;
    private int _mundialMatchDayCount;
    private int _daysSinceSalary;
    private int _nextPlayerId;
    private int _transferDayCount;

    public ManagerState Manager { get; }
    public IReadOnlyList<Club> Clubs => _clubs;
    public CompetitionManager Competition => _competition;
    public IReadOnlyList<SeasonDay> Calendar => _calendar;
    public List<TransferRecord> TransferHistory { get; } = [];
    public List<LoanRecord> ActiveLoans { get; } = [];

    public int CurrentDayIndex => _currentDayIndex;
    public SeasonDay CurrentDay => _calendar[_currentDayIndex];
    public bool IsSeasonOver => _currentDayIndex >= _calendar.Count;

    public GameState(List<Club> clubs, ManagerState manager, int seed)
    {
        _clubs = clubs;
        Manager = manager;
        _baseSeed = seed;
        _competition = new CompetitionManager(clubs, seed);
        _calendar = SeasonCalendar.BuildTemplate();
        _nextPlayerId = YouthAcademy.GetMaxPlayerId(clubs) + 1;
    }

    /// <summary>
    /// Returns the player's club.
    /// </summary>
    public Club PlayerClub => _clubs.First(c => c.Id == Manager.ClubId);

    /// <summary>
    /// Advances one day automatically (no interactive decisions).
    /// </summary>
    public DayResult AdvanceDay()
    {
        if (IsSeasonOver || CareerManager.IsGameOver(Manager) || CareerManager.IsVictory(Manager))
            return new DayResult { Day = CurrentDay, Finished = true };

        var day = CurrentDay;
        var result = new DayResult { Day = day };

        switch (day.Type)
        {
            case DayType.Training:
                ProcessTraining();
                break;

            case DayType.MatchDay:
                ProcessNationalMatchDay(result);
                break;

            case DayType.MundialMatchDay:
                ProcessMundialMatchDay(result);
                break;

            case DayType.TransferWindow:
                ProcessTransferDay(result);
                break;
        }

        FinishDayCommon(result);
        return result;
    }

    /// <summary>
    /// Prepares match day: generates fixtures, simulates all non-player matches.
    /// Returns the player's fixture (if any) and all fixtures for the day.
    /// Use this for interactive mode — call FinishDay() after resolving the player's match.
    /// </summary>
    public MatchDayContext PrepareMatchDay()
    {
        var day = CurrentDay;
        bool isMundial = day.Type == DayType.MundialMatchDay;

        List<MatchFixture> fixtures;
        int daySeed;

        if (isMundial)
        {
            _mundialMatchDayCount++;
            fixtures = _competition.GenerateMundialRound(day.Day);
            daySeed = _baseSeed + day.Day * 1000 + 500;
        }
        else
        {
            _nationalMatchDayCount++;
            fixtures = _competition.GenerateNationalRound(day.Day);
            daySeed = _baseSeed + day.Day * 1000;
        }

        // Find player's fixture
        MatchFixture? playerFixture = fixtures.FirstOrDefault(f =>
            f.HomeClubId == Manager.ClubId || f.AwayClubId == Manager.ClubId);

        // Simulate all non-player fixtures
        var otherFixtures = fixtures.Where(f => f != playerFixture).ToList();
        _competition.SimulateFixtures(otherFixtures, daySeed);

        // Compute the seed for the player's match
        int playerMatchSeed = playerFixture != null
            ? daySeed + fixtures.IndexOf(playerFixture)
            : 0;

        return new MatchDayContext
        {
            AllFixtures = fixtures,
            PlayerFixture = playerFixture,
            PlayerMatchSeed = playerMatchSeed,
            IsMundial = isMundial
        };
    }

    /// <summary>
    /// Finishes a match day after the player's match has been resolved.
    /// Records the player match result, processes economy, career, and advances the day.
    /// </summary>
    public DayResult FinishDay(MatchDayContext ctx, MatchResult? playerResult)
    {
        // Record player match result
        if (ctx.PlayerFixture != null && playerResult != null)
        {
            ctx.PlayerFixture.Result = (
                playerResult.FinalState.ScoreHome,
                playerResult.FinalState.ScoreAway);
        }

        var result = new DayResult { Day = CurrentDay, Fixtures = ctx.AllFixtures };

        // Economy for ALL fixtures
        foreach (var fixture in ctx.AllFixtures)
        {
            var home = _clubs.First(c => c.Id == fixture.HomeClubId);
            var away = _clubs.First(c => c.Id == fixture.AwayClubId);
            var phase = fixture.Phase;
            bool homeWon = fixture.WinnerClubId == home.Id;

            EconomyProcessor.ProcessMatchDay(home, phase, homeWon);
            EconomyProcessor.ProcessMatchDay(away, phase, !homeWon);
        }

        // Advance competition brackets
        if (ctx.IsMundial)
        {
            _competition.AdvanceMundialRound();

            // Update manager reputation
            if (ctx.PlayerFixture != null)
            {
                bool advanced = ctx.PlayerFixture.WinnerClubId == Manager.ClubId;
                CareerManager.UpdateReputation(Manager, ctx.PlayerFixture.Phase, advanced);
            }
        }
        else
        {
            _competition.AdvanceNationalRounds();

            if (_competition.AreNationalsFinished() && _competition.MundialBracket == null)
                _competition.CreateMundial();
        }

        FinishDayCommon(result);
        return result;
    }

    /// <summary>
    /// Returns the MatchConfig for the player's match based on tactical setup.
    /// </summary>
    public MatchConfig BuildPlayerMatchConfig(MatchDayContext ctx, TacticalSetup? tactics)
    {
        if (ctx.PlayerFixture == null)
            throw new InvalidOperationException("No player fixture on this day.");

        var homeClub = _clubs.First(c => c.Id == ctx.PlayerFixture.HomeClubId);
        var awayClub = _clubs.First(c => c.Id == ctx.PlayerFixture.AwayClubId);
        bool isHome = ctx.PlayerFixture.HomeClubId == Manager.ClubId;

        return new MatchConfig
        {
            HomeTeam = homeClub.Team,
            AwayTeam = awayClub.Team,
            Seed = ctx.PlayerMatchSeed,
            HomeTactics = isHome ? tactics : null,
            AwayTactics = isHome ? null : tactics
        };
    }

    private void ProcessTraining()
    {
        foreach (var club in _clubs)
        {
            var updatedPlayers = club.Team.Players.Select(p => p with
            {
                Morale = Math.Min(100, p.Morale + 2)
            }).ToList();

            club.Team = club.Team with { Players = updatedPlayers };
        }
    }

    private void ProcessTransferDay(DayResult result)
    {
        _transferDayCount++;
        int daySeed = _baseSeed + CurrentDay.Day * 2000;
        var rng = new SeededRng(daySeed);

        // AI clubs make transfers
        var aiRecords = AITransferAgent.ProcessDay(_clubs, Manager.ClubId, rng, ref _nextPlayerId);
        TransferHistory.AddRange(aiRecords);
        result.TransferRecords = aiRecords;
    }

    /// <summary>
    /// Returns the next available player ID and increments the counter.
    /// </summary>
    public int GetNextPlayerId(int count = 1)
    {
        int id = _nextPlayerId;
        _nextPlayerId += count;
        return id;
    }

    /// <summary>
    /// Records a player transfer in the history.
    /// </summary>
    public void RecordTransfer(TransferRecord record)
    {
        TransferHistory.Add(record);
    }

    /// <summary>
    /// Records an active loan.
    /// </summary>
    public void RecordLoan(LoanRecord loan)
    {
        ActiveLoans.Add(loan);
    }

    private void ProcessNationalMatchDay(DayResult result)
    {
        _nationalMatchDayCount++;
        var fixtures = _competition.GenerateNationalRound(CurrentDay.Day);
        int daySeed = _baseSeed + CurrentDay.Day * 1000;
        _competition.SimulateFixtures(fixtures, daySeed);
        result.Fixtures = fixtures;

        foreach (var fixture in fixtures)
        {
            var home = _clubs.First(c => c.Id == fixture.HomeClubId);
            var away = _clubs.First(c => c.Id == fixture.AwayClubId);
            var phase = fixture.Phase;
            bool homeWon = fixture.WinnerClubId == home.Id;

            EconomyProcessor.ProcessMatchDay(home, phase, homeWon);
            EconomyProcessor.ProcessMatchDay(away, phase, !homeWon);
        }

        _competition.AdvanceNationalRounds();

        if (_competition.AreNationalsFinished() && _competition.MundialBracket == null)
            _competition.CreateMundial();
    }

    private void ProcessMundialMatchDay(DayResult result)
    {
        _mundialMatchDayCount++;
        var fixtures = _competition.GenerateMundialRound(CurrentDay.Day);
        int daySeed = _baseSeed + CurrentDay.Day * 1000 + 500;
        _competition.SimulateFixtures(fixtures, daySeed);
        result.Fixtures = fixtures;

        foreach (var fixture in fixtures)
        {
            var home = _clubs.First(c => c.Id == fixture.HomeClubId);
            var away = _clubs.First(c => c.Id == fixture.AwayClubId);
            var phase = fixture.Phase;
            bool homeWon = fixture.WinnerClubId == home.Id;

            EconomyProcessor.ProcessMatchDay(home, phase, homeWon);
            EconomyProcessor.ProcessMatchDay(away, phase, !homeWon);
        }

        _competition.AdvanceMundialRound();

        var playerFixture = fixtures.FirstOrDefault(f =>
            f.HomeClubId == Manager.ClubId || f.AwayClubId == Manager.ClubId);
        if (playerFixture != null)
        {
            bool advanced = playerFixture.WinnerClubId == Manager.ClubId;
            CareerManager.UpdateReputation(Manager, playerFixture.Phase, advanced);
        }
    }

    private void FinishDayCommon(DayResult result)
    {
        // Weekly salary (every 7 days)
        _daysSinceSalary++;
        if (_daysSinceSalary >= 7)
        {
            _daysSinceSalary = 0;
            foreach (var club in _clubs)
                EconomyProcessor.ProcessWeeklySalary(club);
            CareerManager.PaySalary(Manager);
        }

        // Check game end conditions
        CareerManager.CheckDismissal(Manager, PlayerClub);
        if (_competition.IsSeasonComplete())
            CareerManager.CheckVictory(Manager, _competition.GetMundialChampion());

        _currentDayIndex++;

        if (CareerManager.IsGameOver(Manager))
            result.GameOver = true;
        if (CareerManager.IsVictory(Manager))
            result.Victory = true;
        if (_currentDayIndex >= _calendar.Count)
            result.Finished = true;
    }
}

/// <summary>
/// Summary of what happened during a single day advance.
/// </summary>
public sealed class DayResult
{
    public required SeasonDay Day { get; init; }
    public IReadOnlyList<MatchFixture> Fixtures { get; set; } = [];
    public IReadOnlyList<TransferRecord> TransferRecords { get; set; } = [];
    public bool GameOver { get; set; }
    public bool Victory { get; set; }
    public bool Finished { get; set; }
}

/// <summary>
/// Context for an interactive match day. Created by PrepareMatchDay(), consumed by FinishDay().
/// </summary>
public sealed class MatchDayContext
{
    public required List<MatchFixture> AllFixtures { get; init; }
    public MatchFixture? PlayerFixture { get; init; }
    public int PlayerMatchSeed { get; init; }
    public bool IsMundial { get; init; }
}

File: engine/game-state.ts
import { DayType } from "@/engine/enums";
import type {
  Club,
  LoanRecord,
  MatchConfig,
  MatchFixture,
  MatchResult,
  ManagerState,
  TacticalSetup,
  TransferRecord,
} from "@/engine/models";
import { winnerClubId } from "@/engine/models";
import { SeededRng } from "@/engine/simulation/rng";
import { CompetitionManager } from "@/engine/competition/competition-manager";
import type { SeasonDay } from "@/engine/competition/season-calendar";
import { buildTemplate } from "@/engine/competition/season-calendar";
import {
  processMatchDay,
  processWeeklySalary,
} from "@/engine/economy/economy-processor";
import { processDay as processAITransferDay } from "@/engine/transfers/ai-transfer-agent";
import { getMaxPlayerId } from "@/engine/transfers/youth-academy";
import {
  updateReputation,
  paySalary,
  checkDismissal,
  checkVictory,
  isGameOver,
  isVictory,
} from "@/engine/career-manager";

/**
 * Summary of what happened during a single day advance.
 */
export interface DayResult {
  day: SeasonDay;
  fixtures: MatchFixture[];
  transferRecords: TransferRecord[];
  gameOver: boolean;
  victory: boolean;
  finished: boolean;
}

/**
 * Context for an interactive match day. Created by prepareMatchDay(), consumed by finishDay().
 */
export interface MatchDayContext {
  allFixtures: MatchFixture[];
  playerFixture: MatchFixture | null;
  playerMatchSeed: number;
  isMundial: boolean;
}

/**
 * Central game state. Tracks the current day, clubs, competitions, and manager.
 * Call advanceDay() for auto-mode, or use prepareMatchDay/finishDay for interactive mode.
 */
export class GameState {
  private readonly _clubs: Club[];
  private _competition: CompetitionManager;
  private _calendar: SeasonDay[];
  private readonly _baseSeed: number;

  private _currentDayIndex: number = 0;
  private _nationalMatchDayCount: number = 0;
  private _mundialMatchDayCount: number = 0;
  private _daysSinceSalary: number = 0;
  private _nextPlayerId: number;
  private _transferDayCount: number = 0;

  readonly manager: ManagerState;
  readonly transferHistory: TransferRecord[] = [];
  readonly activeLoans: LoanRecord[] = [];

  get clubs(): readonly Club[] {
    return this._clubs;
  }

  get competition(): CompetitionManager {
    return this._competition;
  }

  get calendar(): readonly SeasonDay[] {
    return this._calendar;
  }

  get currentDayIndex(): number {
    return this._currentDayIndex;
  }

  get currentDay(): SeasonDay {
    return this._calendar[this._currentDayIndex];
  }

  get isSeasonOver(): boolean {
    return this._currentDayIndex >= this._calendar.length;
  }

  /** Returns the player's club. */
  get playerClub(): Club {
    return this._clubs.find((c) => c.id === this.manager.clubId)!;
  }

  // Exposed for persistence
  get baseSeed(): number {
    return this._baseSeed;
  }
  get nationalMatchDayCount(): number {
    return this._nationalMatchDayCount;
  }
  get mundialMatchDayCount(): number {
    return this._mundialMatchDayCount;
  }
  get daysSinceSalary(): number {
    return this._daysSinceSalary;
  }
  get nextPlayerId(): number {
    return this._nextPlayerId;
  }
  get transferDayCount(): number {
    return this._transferDayCount;
  }

  constructor(clubs: Club[], manager: ManagerState, seed: number) {
    this._clubs = clubs;
    this.manager = manager;
    this._baseSeed = seed;
    this._competition = new CompetitionManager(clubs, seed);
    this._calendar = buildTemplate();
    this._nextPlayerId = getMaxPlayerId(clubs) + 1;
  }

  /**
   * Restores a GameState from saved data. Used by the persistence layer.
   */
  static restore(
    clubs: Club[],
    manager: ManagerState,
    baseSeed: number,
    competition: CompetitionManager,
    calendar: SeasonDay[],
    currentDayIndex: number,
    nationalMatchDayCount: number,
    mundialMatchDayCount: number,
    daysSinceSalary: number,
    nextPlayerId: number,
    transferDayCount: number,
    transferHistory: TransferRecord[],
    activeLoans: LoanRecord[],
  ): GameState {
    const gs = new GameState(clubs, manager, baseSeed);

    // Replace auto-created state with saved state
    gs._currentDayIndex = currentDayIndex;
    gs._nationalMatchDayCount = nationalMatchDayCount;
    gs._mundialMatchDayCount = mundialMatchDayCount;
    gs._daysSinceSalary = daysSinceSalary;
    gs._nextPlayerId = nextPlayerId;
    gs._transferDayCount = transferDayCount;
    gs.transferHistory.push(...transferHistory);
    gs.activeLoans.push(...activeLoans);

    return gs;
  }

  /**
   * Restores using a pre-built CompetitionManager.
   */
  replaceCompetition(competition: CompetitionManager): void {
    this._competition = competition;
  }

  /**
   * Replaces the calendar with saved calendar data.
   */
  replaceCalendar(calendar: SeasonDay[]): void {
    this._calendar = calendar;
  }

  /**
   * Advances one day automatically (no interactive decisions).
   */
  advanceDay(): DayResult {
    if (
      this.isSeasonOver ||
      isGameOver(this.manager) ||
      isVictory(this.manager)
    ) {
      return {
        day: this.currentDay,
        fixtures: [],
        transferRecords: [],
        gameOver: false,
        victory: false,
        finished: true,
      };
    }

    const day = this.currentDay;
    const result: DayResult = {
      day,
      fixtures: [],
      transferRecords: [],
      gameOver: false,
      victory: false,
      finished: false,
    };

    switch (day.type) {
      case DayType.Training:
        this.processTraining();
        break;

      case DayType.MatchDay:
        this.processNationalMatchDay(result);
        break;

      case DayType.MundialMatchDay:
        this.processMundialMatchDay(result);
        break;

      case DayType.TransferWindow:
        this.processTransferDay(result);
        break;
    }

    this.finishDayCommon(result);
    return result;
  }

  /**
   * Returns true if the player's club is still alive in the current competition phase.
   * Use this to decide whether to show "Play Match" or "Advance Day" in the UI.
   */
  isPlayerInCurrentCompetition(): boolean {
    const day = this.currentDay;
    if (day.type === DayType.MatchDay) {
      return this._competition.isTeamInNationals(this.manager.clubId);
    }
    if (day.type === DayType.MundialMatchDay) {
      const mb = this._competition.mundialBracket;
      return mb !== null && mb.hasTeam(this.manager.clubId);
    }
    return false;
  }

  /**
   * Prepares match day: generates fixtures, simulates all non-player matches.
   * Returns the player's fixture (if any) and all fixtures for the day.
   * Use this for interactive mode -- call finishDay() after resolving the player's match.
   */
  prepareMatchDay(): MatchDayContext {
    const day = this.currentDay;
    const isMundial = day.type === DayType.MundialMatchDay;

    let fixtures: MatchFixture[];
    let daySeed: number;

    if (isMundial) {
      this._mundialMatchDayCount++;
      fixtures = this._competition.generateMundialRound(day.day);
      daySeed = this._baseSeed + day.day * 1000 + 500;
    } else {
      this._nationalMatchDayCount++;
      fixtures = this._competition.generateNationalRound(day.day);
      daySeed = this._baseSeed + day.day * 1000;
    }

    // Find player's fixture
    const playerFixture =
      fixtures.find(
        (f) =>
          f.homeClubId === this.manager.clubId ||
          f.awayClubId === this.manager.clubId,
      ) ?? null;

    // Simulate all non-player fixtures (or all fixtures if player is eliminated)
    const otherFixtures = fixtures.filter((f) => f !== playerFixture);
    this._competition.simulateFixtures(otherFixtures, daySeed);

    // Compute the seed for the player's match
    const playerMatchSeed =
      playerFixture !== null
        ? daySeed + fixtures.indexOf(playerFixture)
        : 0;

    return {
      allFixtures: fixtures,
      playerFixture,
      playerMatchSeed,
      isMundial,
    };
  }

  /**
   * Finishes a match day after the player's match has been resolved.
   * Records the player match result, processes economy, career, and advances the day.
   */
  finishDay(
    ctx: MatchDayContext,
    playerResult: MatchResult | null,
  ): DayResult {
    // Record player match result
    if (ctx.playerFixture !== null && playerResult !== null) {
      ctx.playerFixture.result = [
        playerResult.finalState.scoreHome,
        playerResult.finalState.scoreAway,
      ];
    }

    const result: DayResult = {
      day: this.currentDay,
      fixtures: ctx.allFixtures,
      transferRecords: [],
      gameOver: false,
      victory: false,
      finished: false,
    };

    // Economy for ALL fixtures
    for (const fixture of ctx.allFixtures) {
      const home = this._clubs.find(
        (c) => c.id === fixture.homeClubId,
      )!;
      const away = this._clubs.find(
        (c) => c.id === fixture.awayClubId,
      )!;
      const phase = fixture.phase;
      const homeWon = winnerClubId(fixture) === home.id;

      processMatchDay(home, phase, homeWon);
      processMatchDay(away, phase, !homeWon);
    }

    // Advance competition brackets
    if (ctx.isMundial) {
      this._competition.advanceMundialRound();

      // Update manager reputation
      if (ctx.playerFixture !== null) {
        const advanced =
          winnerClubId(ctx.playerFixture) === this.manager.clubId;
        updateReputation(
          this.manager,
          ctx.playerFixture.phase,
          advanced,
        );
      }
    } else {
      this._competition.advanceNationalRounds();

      if (
        this._competition.areNationalsFinished() &&
        this._competition.mundialBracket === null
      ) {
        this._competition.createMundial();
      }
    }

    this.finishDayCommon(result);
    return result;
  }

  /**
   * Returns the MatchConfig for the player's match based on tactical setup.
   */
  buildPlayerMatchConfig(
    ctx: MatchDayContext,
    tactics: TacticalSetup | null,
  ): MatchConfig {
    if (ctx.playerFixture === null) {
      throw new Error("No player fixture on this day.");
    }

    const homeClub = this._clubs.find(
      (c) => c.id === ctx.playerFixture!.homeClubId,
    )!;
    const awayClub = this._clubs.find(
      (c) => c.id === ctx.playerFixture!.awayClubId,
    )!;
    const isHome =
      ctx.playerFixture.homeClubId === this.manager.clubId;

    return {
      homeTeam: homeClub.team,
      awayTeam: awayClub.team,
      seed: ctx.playerMatchSeed,
      homeTactics: isHome ? tactics : null,
      awayTactics: isHome ? null : tactics,
      homeSubstitutions: null,
      awaySubstitutions: null,
      homeCard: null,
      awayCard: null,
    };
  }

  /**
   * Returns the next available player ID and increments the counter.
   */
  getNextPlayerId(count: number = 1): number {
    const id = this._nextPlayerId;
    this._nextPlayerId += count;
    return id;
  }

  /**
   * Records a player transfer in the history.
   */
  recordTransfer(record: TransferRecord): void {
    this.transferHistory.push(record);
  }

  /**
   * Records an active loan.
   */
  recordLoan(loan: LoanRecord): void {
    this.activeLoans.push(loan);
  }

  private processTraining(): void {
    for (const club of this._clubs) {
      const updatedPlayers = club.team.players.map((p) => ({
        ...p,
        morale: Math.min(100, p.morale + 2),
      }));

      club.team = {
        ...club.team,
        players: updatedPlayers,
      };
    }
  }

  private processTransferDay(result: DayResult): void {
    this._transferDayCount++;
    const daySeed = this._baseSeed + this.currentDay.day * 2000;
    const rng = new SeededRng(daySeed);

    // AI clubs make transfers
    const aiRecords = processAITransferDay(
      this._clubs,
      this.manager.clubId,
      rng,
    );
    this.transferHistory.push(...aiRecords);
    result.transferRecords = aiRecords;
  }

  private processNationalMatchDay(result: DayResult): void {
    this._nationalMatchDayCount++;
    const fixtures = this._competition.generateNationalRound(
      this.currentDay.day,
    );
    const daySeed = this._baseSeed + this.currentDay.day * 1000;
    this._competition.simulateFixtures(fixtures, daySeed);
    result.fixtures = fixtures;

    for (const fixture of fixtures) {
      const home = this._clubs.find(
        (c) => c.id === fixture.homeClubId,
      )!;
      const away = this._clubs.find(
        (c) => c.id === fixture.awayClubId,
      )!;
      const phase = fixture.phase;
      const homeWon = winnerClubId(fixture) === home.id;

      processMatchDay(home, phase, homeWon);
      processMatchDay(away, phase, !homeWon);
    }

    this._competition.advanceNationalRounds();

    if (
      this._competition.areNationalsFinished() &&
      this._competition.mundialBracket === null
    ) {
      this._competition.createMundial();
    }
  }

  private processMundialMatchDay(result: DayResult): void {
    this._mundialMatchDayCount++;
    const fixtures = this._competition.generateMundialRound(
      this.currentDay.day,
    );
    const daySeed =
      this._baseSeed + this.currentDay.day * 1000 + 500;
    this._competition.simulateFixtures(fixtures, daySeed);
    result.fixtures = fixtures;

    for (const fixture of fixtures) {
      const home = this._clubs.find(
        (c) => c.id === fixture.homeClubId,
      )!;
      const away = this._clubs.find(
        (c) => c.id === fixture.awayClubId,
      )!;
      const phase = fixture.phase;
      const homeWon = winnerClubId(fixture) === home.id;

      processMatchDay(home, phase, homeWon);
      processMatchDay(away, phase, !homeWon);
    }

    this._competition.advanceMundialRound();

    const playerFixture = fixtures.find(
      (f) =>
        f.homeClubId === this.manager.clubId ||
        f.awayClubId === this.manager.clubId,
    );
    if (playerFixture != null) {
      const advanced =
        winnerClubId(playerFixture) === this.manager.clubId;
      updateReputation(this.manager, playerFixture.phase, advanced);
    }
  }

  private finishDayCommon(result: DayResult): void {
    // Weekly salary (every 7 days)
    this._daysSinceSalary++;
    if (this._daysSinceSalary >= 7) {
      this._daysSinceSalary = 0;
      for (const club of this._clubs) {
        processWeeklySalary(club);
      }
      paySalary(this.manager);
    }

    // Check game end conditions
    checkDismissal(this.manager, this.playerClub);
    if (this._competition.isSeasonComplete()) {
      checkVictory(this.manager, this._competition.getMundialChampion());
    }

    this._currentDayIndex++;

    if (isGameOver(this.manager)) {
      result.gameOver = true;
    }
    if (isVictory(this.manager)) {
      result.victory = true;
    }
    if (this._currentDayIndex >= this._calendar.length) {
      result.finished = true;
    }
  }
}
